src/twisted/internet/tcp.py

```diff
@@ -100,6 +100,39 @@ class _TLSServerMixin(object):
 _portNameType = (str, unicode)
 
 
+def _getrealname(addr):
+    """
+    Return a 2-tuple of socket IP and port for IPv4 and a 4-tuple of
+    socket IP, port, flowInfo, and scopeID.  With IPv6 addresses, this
+    preserves the interface portion.
+
+    @param addr: A 2-tuple for IPv4 information or a 4-tuple for IPv6
+        information.
+    """
+    if len(addr) == 4:
+        # IPv6
+        host = socket.getnameinfo(addr, socket.NI_NUMERICHOST)[0]
```

Comment:
I think you really want `socket.NI_NUMERICHOST | socket.NI_NUMERICSERV`; you don't need to do service-name lookup, so this is asking the OS to do some pointless I/O on a part of the address you'll immediately throw away.